### ISSUE TITLE
[Feature request] Depth Smooth Loss

### ISSUE THREAD

**Issue by ClementPinard** [kornia/kornia#60]
Hello, I happened to have worked on depth smoothness,
and even if when they are used they are called depth smooth, the value that is smoothed is actually the inverse depth. So maybe we should add it to the right terminology, otherwise it might get confusing and people will try to lower spatial gradient of depth instead of inverse depth.

The loss is taken from https://github.com/tensorflow/models/blob/master/research/struct2depth/model.py#L625-L641

But you can see here, that they apply it on "disparity" (calling it inverse depth is better since disparity is only valid for stereo vision)
https://github.com/tensorflow/models/blob/master/research/struct2depth/model.py#L464-L465

A quick example on why it is important not to confuse the two is the scene with inifinite horizontal plane, where depth goes to infinity (along with its gradient).

http://perso.ensta-paristech.fr/~pinard/images/plane.png
http://perso.ensta-paristech.fr/~pinard/images/depth_vs_disp.png

Another point to discuss is the fact that depth smoothness can also be applied in  a 2nd order fashion, eg as here : https://github.com/tinghuiz/SfMLearner/blob/master/SfMLearner.py#L156-L167

These loss are very close the Total Variation loss, so maybe we can call them TVloss and TVVloss.

More generally, I'm currently writing a report on depth smoothness constraint for autosupervised depth training, I'll keep you updated on it, but the general idea is that it's probably better to try to minimize the square of gradients to have a something similar to diffusion operation, the same L2 weight decay is better than L1 weight decay.

**Comment by edgarriba** [comment]
HI @ClementPinard ! thanks for the feedback.

Regarding the terminology I agree and I'll make the changes ASAP.

As for the second order approach, I think the easier could be adding a flag to the signature. In the end it only adds the second order to the returned loss.

About the ``TotalVariationLoss``, it only accepts a single tensor which is the depth, right ? In the ``DepthSmoothnessLoss`` the loss is weighted as well by the image gradients. So, I see different implementations here. What do you think ?

**Comment by ClementPinard** [comment]
Well it's not TVLoss per se, but you could see extended TV loss as a weighted sum of gradients. For example, it's called TV loss for DTAM : https://www.robots.ox.ac.uk/~vgg/rg/papers/newcombe_davison__2011__dtam.pdf which i think is a pretty solid reference, and it's probably where the exponential weight of image gradient comes from originally.

The regular TVLoss would then be the special case of the same weight everywhere.
The best way to see it would be to define an extended TV loss with a supplementary parameter kappa with the weights being exp(-grad(I)/kappa). so with kappa -> infinity, we get the regular TVloss, and kappa = 1 we get today's depth smoothness loss.
kappa is then a paramter of how influenced to image edges the smoothing is.

And why we have the inverse relation kappa -> infinity  = no influence is because originally, this comes from perona malik filter. kappa was the in the context of anisotropic diffusion (so adding the square of inverse depth gradient instead of absolute value) the gradient threshold at which edges were reinforced or smoothed.

The link is not easily gometrically interpretable now, but you can see it as a "nominal image gradient" to which the weight will be normalized.

**Comment by edgarriba** [comment]
@ClementPinard are there any specific actions that could be taken here or any improvement in the API, add extra missing loss to this ? Otherwise we can leave this closed.
